packages/vote/fpc/fpc.go

```diff
@@ -0,0 +1,267 @@
+type FPC struct {
+	events           vote.Events
+	opinionGiverFunc vote.OpinionGiverFunc
+	// the lifo queue of newly enqueued items to vote on.
+	queue *list.List
+	// contains a set of currently queued items.
+	queueSet map[string]struct{}
+	queueMu  sync.Mutex
+	// contains the set of current vote contexts.
+	ctxs   map[string]*VoteContext
+	ctxsMu sync.RWMutex
+	// parameters to use within FPC.
+	paras *Parameters
+	// indicates whether the last round was performed successfully
+	lastRoundCompletedSuccessfully bool
+	rng                            *rand.Rand
+}
+
+func (f *FPC) Vote(id string, initOpn vote.Opinion) error {
+	f.queueMu.Lock()
+	defer f.queueMu.Unlock()
+	f.ctxsMu.RLock()
+	defer f.ctxsMu.RUnlock()
+	if _, alreadyQueued := f.queueSet[id]; alreadyQueued {
+		return fmt.Errorf("%w: %s", ErrVoteAlreadyOngoing, id)
+	}
+	if _, alreadyOngoing := f.ctxs[id]; alreadyOngoing {
+		return fmt.Errorf("%w: %s", ErrVoteAlreadyOngoing, id)
+	}
+	f.queue.PushBack(newVoteContext(id, initOpn))
+	f.queueSet[id] = struct{}{}
+	return nil
+}
+
+func (f *FPC) IntermediateOpinion(id string) (vote.Opinion, error) {
+	f.ctxsMu.RLock()
+	defer f.ctxsMu.RUnlock()
+	voteCtx, has := f.ctxs[id]
+	if !has {
+		return vote.Unknown, fmt.Errorf("%w: %s", vote.ErrVotingNotFound, id)
+	}
+	return voteCtx.LastOpinion(), nil
+}
+
+func (f *FPC) Events() vote.Events {
+	return f.events
+}
+
+// Round enqueues new items, sets opinions on active vote contexts, finalizes them and then
+// queries for opinions.
+func (f *FPC) Round(rand float64) error {
+	// enqueue new voting contexts
+	f.enqueue()
+	// we can only form opinions when the last round was actually executed successfully
+	if f.lastRoundCompletedSuccessfully {
+		// form opinions by using the random number supplied for this new round
+		f.formOpinions(rand)
+		// clean opinions on vote contexts where an opinion was reached in FinalizationThreshold
+		// number of rounds and clear those who failed to be finalized in MaxRoundsPerVoteContext.
+		f.finalizeOpinions()
+	}
+	// query for opinions on the current vote contexts
+	err := f.queryOpinions()
+	f.lastRoundCompletedSuccessfully = err == nil
+	return err
+}
+
+// enqueues items for voting
+func (f *FPC) enqueue() {
+	f.queueMu.Lock()
+	defer f.queueMu.Unlock()
+	f.ctxsMu.Lock()
+	defer f.ctxsMu.Unlock()
+	for ele := f.queue.Front(); ele != nil; ele = f.queue.Front() {
+		voteCtx := ele.Value.(*VoteContext)
+		f.ctxs[voteCtx.id] = voteCtx
+		f.queue.Remove(ele)
+		delete(f.queueSet, voteCtx.id)
+	}
+}
+
+// formOpinions updates the opinion for ongoing vote contexts by comparing their liked percentage
+// against the threshold appropriate for their given rounds.
+func (f *FPC) formOpinions(rand float64) {
+	f.ctxsMu.RLock()
+	defer f.ctxsMu.RUnlock()
+	for _, voteCtx := range f.ctxs {
+		// when the vote context is new there's no opinion to form
+		if voteCtx.IsNew() {
+			continue
+		}
+
+		lowerThreshold := f.paras.SubsequentRoundsLowerBoundThreshold
```

Comment:
Let's fix the `lowerThreshold` to `0.5` so that the range for the subsequent rounds is always between 0.5 and 1-Beta. I know it's different from the FPC white paper [beta - 1-beta] but simulations have shown that FPC performs better in this way.